### PR TITLE
Add iSCSI ReplacementTimeout  API's to keep volume in RW state during "Node Down" scenario

### PR DESCRIPTION
**Pull request by mittachaitu**
**What this PR does / why we need it**:
 Add ReplacementTimeout field in ISCSIVolumeSource and
 ISCSIPersistentVolumeSource spec as API changes.
 ReplacementTimeout controls how long the iSCSI layer should wait
 for a timed-out path/session to re-establish itself before failing
 any commands on it.

Below one is the example to pass a ReplacementTimeout setting parameter:
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: iscsi-pv
spec:
  storageClassName: standard
  capacity:
    storage: 1Gi
  accessModes:
    - ReadWriteOnce
  iscsi:
     targetPortal: 192.68.0.134:3260
     iqn: iqn.2018-09.com.storage.volume:vol1
     lun: 0
     fsType: 'ext4'
     readOnly: false
     replacementTimeout: 300
 ```

cc /sig storage
```release-note
NONE
```

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

### CODE REVIEW

**Comment by prateekpandey14** [comment]
cc @rootfs

**Comment by jsafrane** [comment]
> ```
> provisioner: kubernetes.io/storageos
> parameters:
>   storage.alpha.kubernetes.io/iscsiTimeout: "node.session.timeo.replacement_timeout=350"
> ```

That does not make sense, `kubernetes.io/storageos` provisioner will create StorageOS volumes, not iscsi.

File: pkg/volume/iscsi/iscsi.go [hunk 1]
```diff
@@ -53,6 +53,18 @@ const (
...
 )
 
+//iscsiTimeoutKey is used to update the iscsi Timeout setting i.e
+//(node.session.timeo.replacement_timeout) to keep volume in RW state during node down cases
+type iscsiTimeoutKey string
+type iscsiTimeout struct {
+	Name  iscsiTimeoutKey
+	Value string
+}
+type iscsiTimeouts []iscsiTimeout
```
Comment (jsafrane) [inline pkg/volume/iscsi/iscsi.go]: Why is it array? There  can be only one annotation in a PV.

File: pkg/volume/iscsi/iscsi.go [hunk 2]
```diff
@@ -506,6 +519,26 @@ func getISCSISecretNameAndNamespace(spec *volume.Spec, defaultSecretNamespace st
 	return "", "", fmt.Errorf("Spec does not reference an ISCSI volume type")
 }
 
+// updateFromPV will read iSCSI Timeout settings from Persistent Volume object
+func (kvp iscsiTimeouts) getISCSITimeoutSettingInfo(spec *volume.Spec) iscsiTimeouts {
```
Comment (jsafrane) [inline pkg/volume/iscsi/iscsi.go]: Please rewrite to `func getISCSITimeoutSettingInfo(...) iscsiTimeouts`, there is no need for this to be method of `iscsiTimeouts`

File: pkg/volume/iscsi/iscsi_util.go [hunk 3]
```diff
@@ -86,10 +88,34 @@ func updateISCSIDiscoverydb(b iscsiDiskMounter, tp string) error {
...
+func updateISCSINodeTimeouts(b iscsiDiskMounter, tp string) {
+	//Updating the iSCSI timeouts settings to keep volume in RW state in the node
+	//down scenario i.e when the node goes down K8s take 5min to schedule the node
+	//until that time we can keep volume in RW state by updating these timeouts
+	//setting value to 300+ seconds. This updation will execute if PV object
+	//contains a valid key, value pair
+	if b.iscsiTimeouts != nil {
+		for _, iscsiSetting := range b.iscsiTimeouts {
+			if iscsiSetting.Name != "" && iscsiSetting.Value != "" {
+				out, err := b.exec.Run("iscsiadm", "-m", "node", "-p", tp, "-T", b.Iqn, "-I", b.Iface, "-o", "update", "-n", string(iscsiSetting.Name), "-v", iscsiSetting.Value)
```
Comment (jsafrane) [inline pkg/volume/iscsi/iscsi_util.go]: This just blindly sets **any** parameters from the annotation to `iscsiadm`. It does not check if it's timeout related. This should be validated somewhere and you should allow only very specific keys.

**Comment by rootfs** [comment]
iscsi timeout can be set in the default iscsid.conf (on rhel/centos/fedora it is /etc/iscsi/iscsid.conf) on the nodes

**Comment by mittachaitu** [comment]
> > ```
> > provisioner: kubernetes.io/storageos
> > parameters:
> >   storage.alpha.kubernetes.io/iscsiTimeout: "node.session.timeo.replacement_timeout=350"
> > ```
> 
> That does not make sense, `kubernetes.io/storageos` provisioner will create StorageOS volumes, not iscsi.

It's just an example, anyways updated to `kubernetes.io/provisioner_iscsi`

File: pkg/volume/iscsi/iscsi.go [hunk 4]
```diff
@@ -53,6 +53,18 @@ const (
...
+//(node.session.timeo.replacement_timeout) to keep volume in RW state during node down cases
+type iscsiTimeoutKey string
+type iscsiTimeout struct {
+	Name  iscsiTimeoutKey
+	Value string
+}
+type iscsiTimeouts []iscsiTimeout
+
+//iscsiAnnotationLblKey is used to read timeout setting and value from PV object
+const iscsiAnnotationLblKey = "storage.alpha.kubernetes.io/iscsiTimeout"
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/volume/iscsi/iscsi.go]: I was thinking the same, should we add `TimeOut` API field in `ISCSIVolumeSource` and validate the same ?

**Comment by mittachaitu** [comment]
> iscsi timeout can be set in the default iscsid.conf (on rhel/centos/fedora it is /etc/iscsi/iscsid.conf) on the node

**Comment by mittachaitu** [comment]
@jsafrane added the ReplacementTimeout as proper API field instead of annotation, PTAL

File: pkg/apis/core/validation/validation_test.go [hunk 5]
```diff
@@ -10285,14 +10353,14 @@ func TestValidateNode(t *testing.T) {
...
 			expectedFields := map[string]bool{
-				"metadata.name":         true,
-				"metadata.labels":       true,
-				"metadata.annotations":  true,
-				"metadata.namespace":    true,
-				"spec.externalID":       true,
-				"spec.taints[0].key":    true,
-				"spec.taints[0].value":  true,
-				"spec.taints[0].effect": true,
+				"metadata.name":                                                                                               true,
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation_test.go]: gofmt changes should not be part of this PR. Do you use go 1.11?

File: pkg/apis/core/types.go [hunk 6]
```diff
@@ -650,6 +650,13 @@ type ISCSIVolumeSource struct {
 	// <target portal>:<volume name> will be created for the connection.
 	// +optional
 	InitiatorName *string
+	// iSCSI Replacement Timeout field required to set the iSCSI TimeoutSetting to wait for session re-establishment
+	// before failing SCSI commands back to the application when running the Linux SCSI Layer error handler.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
+	// is logged back in, or until the user runs the logout command.
+	// +optional
+	ReplacementTimeout string
```
Comment (jsafrane) [inline pkg/apis/core/types.go]: It should be `*int32` (or `*int64`) - the field needs to be optional and nil value should mean the behavior we have today.

File: pkg/apis/core/types.go [hunk 7]
```diff
@@ -650,6 +650,13 @@ type ISCSIVolumeSource struct {
 	// <target portal>:<volume name> will be created for the connection.
 	// +optional
 	InitiatorName *string
+	// iSCSI Replacement Timeout field required to set the iSCSI TimeoutSetting to wait for session re-establishment
+	// before failing SCSI commands back to the application when running the Linux SCSI Layer error handler.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
```
Comment (jsafrane) [inline pkg/apis/core/types.go]: You validate that the number cannot be negative and this comment allows negative value.

File: pkg/apis/core/types.go [hunk 8]
```diff
@@ -650,6 +650,13 @@ type ISCSIVolumeSource struct {
 	// <target portal>:<volume name> will be created for the connection.
 	// +optional
 	InitiatorName *string
+	// iSCSI Replacement Timeout field required to set the iSCSI TimeoutSetting to wait for session re-establishment
```
Comment (jsafrane) [inline pkg/apis/core/types.go]: Better "ReplacementTimeout controls how long the iSCSI layer should wait for a timed-out path/session to reestablish itself before failing any commands on it."

File: pkg/apis/core/types.go [hunk 9]
```diff
@@ -650,6 +650,13 @@ type ISCSIVolumeSource struct {
...
+	// before failing SCSI commands back to the application when running the Linux SCSI Layer error handler.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
```
Comment (jsafrane) [inline pkg/apis/core/types.go]: You should also describe meaning of empty value.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 10]
```diff
@@ -1156,6 +1156,13 @@ type ISCSIVolumeSource struct {
 	// <target portal>:<volume name> will be created for the connection.
 	// +optional
 	InitiatorName *string `json:"initiatorName,omitempty" protobuf:"bytes,12,opt,name=initiatorName"`
+	// iSCSI Replacement Timeout field required to set the iSCSI TimeoutSetting to wait for session re-establishment
+	// before failing SCSI commands back to the application when running the Linux SCSI Layer error handler.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
+	// is logged back in, or until the user runs the logout command.
+	// +optional
+	ReplacementTimeout string `json:"replacementTimeout,omitempty" portobuf:"bytes,13,opt,name=replacementTimeout" protobuf:"varint,13,opt,name=replacementTimeout"`
```
Comment (jsafrane) [inline staging/src/k8s.io/api/core/v1/types.go]: remove `portobuf`.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 11]
```diff
@@ -1202,6 +1209,13 @@ type ISCSIPersistentVolumeSource struct {
 	// <target portal>:<volume name> will be created for the connection.
 	// +optional
 	InitiatorName *string `json:"initiatorName,omitempty" protobuf:"bytes,12,opt,name=initiatorName"`
+	// iSCSI Replacement Timeout field required to set the iSCSI TimeoutSetting to wait for session re-establishment
+	// before failing SCSI commands back to the application when running the Linux SCSI Layer error handler.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
+	// is logged back in, or until the user runs the logout command.
+	// +optional
+	ReplacementTimeout string `json:"replacementTimeout,omitempty" portobuf:"bytes,13,opt,name=replacementTimeout" protobuf:"bytes,13,opt,name=replacementTimeout"`
```
Comment (jsafrane) [inline staging/src/k8s.io/api/core/v1/types.go]: remove `portobuf`.

**Review by jsafrane — COMMENTED** [review]
Now it looks much better.

Please remove the first commit and do separate commits for adding the new field (+ its handling) and the generated stuff.

**Comment by jsafrane** [comment]
cc @j-griffith @bswartz @redbaron for iSCSI expertise. What do you think about this PR?

**Comment by bswartz** [comment]
No objections from me. I don't play around with iSCSI timeout settings personally. It's worth noting that enhancements like this will need to be reimplemented in CSI plugins and/or the common iSCSI library, or they will be lost in the migration to the out-of-tree iSCSI CSI plugin.

File: pkg/apis/core/validation/validation_test.go (hunk 5 above)
Comment (mittachaitu) [reply pkg/apis/core/validation/validation_test.go]: Yes, I am using  the same go version

File: pkg/apis/core/types.go (hunk 6 above)
Comment (prateekpandey14) [reply pkg/apis/core/types.go]: @jsafrane just to add what @mittachaitu said , the other thing is if we use integer value, we have to convert it to string before passing it to UpdateISCSI method as iscsi-update operator accepts string args as a value, so anyway we have to do type conversion.

Comment (jsafrane) [reply pkg/apis/core/types.go]: This API is targeted to users, not iscsiadm. Users must see that an integer is expected. You'll get validation for free. Using OpenAPI, the validation can be even done on the client side.

How is the parameter converted internally and passed to iscsiadm is implementation detail.

Comment (prateekpandey14) [reply pkg/apis/core/types.go]: Agreed what you said, thanks,  we have changed to `*int32` and testing the changes right now.

**Comment by prateekpandey14** [comment]
@jsafrane requested changes are done PTAL

File: pkg/apis/core/validation/validation.go [hunk 12]
```diff
@@ -707,6 +708,12 @@ func validateISCSIVolumeSource(iscsi *core.ISCSIVolumeSource, fldPath *field.Pat
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("initiatorname"), initiator, "must be valid format"))
 		}
 	}
+	if iscsi.ReplacementTimeout != nil {
+		timeout := strconv.Itoa(int(*iscsi.ReplacementTimeout))
+		if len(timeout) == 0 {
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation.go]: this can't be ever true, `Itoa` never returns empty string

File: staging/src/k8s.io/api/core/v1/types.go [hunk 13]
```diff
@@ -1156,6 +1156,15 @@ type ISCSIVolumeSource struct {
...
 	InitiatorName *string `json:"initiatorName,omitempty" protobuf:"bytes,12,opt,name=initiatorName"`
+	// ReplacementTimeout controls how long the iSCSI layer should wait for a
+	// timed-out path/session to reestablish itself before failing any commands on it.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
+	// is logged back in, or until the user runs the logout command.
+	// If the ReplacementTimeout field is not provided or empty then it remains with
+	// default value i.e 120 seconds.
+	// +optional
+	ReplacementTimeout *int32 `json:"replacementTimeout,omitempty"`
```
Comment (jsafrane) [inline staging/src/k8s.io/api/core/v1/types.go]: Please add protobuf tags.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 14]
```diff
@@ -1202,6 +1211,15 @@ type ISCSIPersistentVolumeSource struct {
...
```
Comment (jsafrane) [inline staging/src/k8s.io/api/core/v1/types.go]: Please add protobuf tags.

File: pkg/apis/core/types.go [hunk 15]
```diff
@@ -698,6 +707,15 @@ type ISCSIPersistentVolumeSource struct {
 	// <target portal>:<volume name> will be created for the connection.
 	// +optional
 	InitiatorName *string
+	// ReplacementTimeout controls how long the iSCSI layer should wait for a
+	// timed-out path/session to reestablish itself before failing any commands on it.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
```
Comment (jsafrane) [inline pkg/apis/core/types.go]: https://github.com/open-iscsi/open-iscsi does not describe meaning of negative value. How do you know that negative value means basically no timeout?

File: pkg/volume/iscsi/iscsi_util.go [hunk 16]
```diff
@@ -86,10 +91,29 @@ func updateISCSIDiscoverydb(b iscsiDiskMounter, tp string) error {
...
 		}
 	}
+
 	return nil
 }
 
+//updateISCSINodeTimeouts will update the timeout settings with the
+//values provided in ISCSIVolumeSource and ISCSIPersistentVolumeSource spec
+func updateISCSINodeTimeouts(b iscsiDiskMounter, tp string) {
+	//Updating the iSCSI timeouts settings to keep volume in RW state in the node
```
Comment (jsafrane) [inline pkg/volume/iscsi/iscsi_util.go]: I think you can remove the comments and add them to documentation of iSCSI driver. It's intended to user and not Kubernetes developer, right?

File: pkg/volume/iscsi/iscsi.go [hunk 17]
```diff
@@ -506,12 +507,23 @@ func getISCSISecretNameAndNamespace(spec *volume.Spec, defaultSecretNamespace st
 	return "", "", fmt.Errorf("Spec does not reference an ISCSI volume type")
 }
 
+// updateFromPV will read iSCSI Timeout settings from iSCSI VolumeSource or PersistentVolumeSource
+func getISCSITimeoutSettingInfo(spec *volume.Spec) string {
```
Comment (jsafrane) [inline pkg/volume/iscsi/iscsi.go]: Function name and the comment right above are not in sync.

File: pkg/apis/core/types.go (hunk 15 above)
Comment (mittachaitu) [reply pkg/apis/core/types.go]: @jsafrane https://github.com/open-iscsi/open-iscsi/blob/master/etc/iscsid.conf in Timeout section described about the negative value of replacement timeout.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 14 above)
Comment (mittachaitu) [reply staging/src/k8s.io/api/core/v1/types.go]: ```protobuf``` tag is added in Auto-generated commit.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 13 above)
Comment (mittachaitu) [reply staging/src/k8s.io/api/core/v1/types.go]: protobuf tags are automatically added in autogenerated files i.e in the second commit.

**Comment by mittachaitu** [comment]
/ test pull-kubernetes-e2e-gce-100-performance

**Comment by prateekpandey14** [comment]
@jsafrane @saad-ali , as suggested changes are done , can you please approve this PR.

**Comment by humblec** [comment]
I have a high level question/thought  on this PR.  That said,  there are many configuration options available for iscsi which is available through its `conf` file. Bringing these parameters to ISCSI spec may not be a good idea. Then what can be brought in is also a question.  IMO, we should consider `target` specific configuration and not really the deamon specific configuration to the spec.  afaict `ReplacementTimeout` does not fall into this  ( `target`) category . It could be configured by admin via conf changes if they need to tweak the default values. There are `many timeout` values available, bringing those to the spec wont be a good idea I believe.  

@jsafrane @saad-ali @rootfs thoughts ?

**Comment by mittachaitu** [comment]
> I have a high level question/thought on this PR. That said, there are many configuration options available for iscsi which is available through its `conf` file. Bringing these parameters to ISCSI spec may not be a good idea. Then what can be brought in is also a question. IMO, we should consider `target` specific configuration and not really the deamon specific configuration to the spec. afaict `ReplacementTimeout` does not fall into this ( `target`) category . It could be configured by admin via conf changes if they need to tweak the default values. There are `many timeout` values available, bringing those to the spec wont be a good idea I believe.
> 
> @jsafrane @saad-ali @rootfs thoughts ?

thanks @humblec for your thoughts.
In non k8s environment, the user does iscsi login, iscsi update command (based on his need), does the mount and uses the mountpoint in the application.
In k8s, iSCSI login will be done by kubelet, and by default, settings in iscsi conf file settings will be applied to the session. Any need to modify these settings need re-login which means unmount of volume and restart of the application.
As @bswartz mentioned, this enhancement is worth to be considered for CSI as well.
Regarding your other point, IMO, we should consider the settings which need re-login.

**Comment by jsafrane** [comment]
I think that @humblec raised valid points. We can't accommodate every iSCSI option into kubernetes API. And there is a way how to configure the timeout in a config file. Should we document how to change the config instead?
